### ISSUE TITLE
Log barrier loss implementation

### ISSUE THREAD

**Issue by paumarquez** [LIVIAETS/boxes_tightness_prior#5]
Hello, I'm doing the final thesis of my bachelor's degree about object detection and I'm trying to apply your method to get a mask instead of the bounding boxes.

I've been extracting the key parts of your code (bounds and losses) to adapt it to my code but there is something I don't understand.

When applying the LogBarrierLoss, the bounds are given by parameter, as they are calculated when loading the data from the dataloader for the whole image (the positive area for the whole image times the margins).

However, the inference is done for each slice of the image (I guess it's because of memory issues, otherwise I'd appreciate to know why). When applying this loss to the prediction, the predicted mask for that slice is compared with the bounds of the full image from what I understand, instead of the bounds for that slice.
https://github.com/LIVIAETS/boxes_tightness_prior/blob/a1356a2072a2d16be7635d395b43e518421c8561/main.py#L168

What am I missing?

Thanks in advance and great job! It's being very helpful.

**Comment by HKervadec** [comment]
Hi Pau, and thanks for the kind words.

The code is rather complex (perhaps a bit too much), as I use that general framework for most of my research and I keep adding things while trying to maintain backward compatiblity.

> When applying the LogBarrierLoss, the bounds are given by parameter, as they are calculated when loading the data from the dataloader for the whole image (the positive area for the whole image times the margins).

Indeed. The Logbarrier, at its core, needs only three inputs: softmax predictions, the function to compute, and the bounds of that function. The bounds are indeed generated at the dataloader level, usings functions defined in `bounds.py`.  What I do is much more generic, as (over the years) I had to deal with various settings.

> However, the inference is done for each slice of the image (I guess it's because of memory issues, otherwise I'd appreciate to know why). When applying this loss to the prediction, the predicted mask for that slice is compared with the bounds of the full image from what I understand, instead of the bounds for that slice.

I am not sure I understood what you meant here, but I guess I can already explicit a bit more the differences between my training and validation loops. 

In validation, the loss and bounds are computed the same way as training—only  to monitor things and (potentially) avoid overfitting. (Though this is less true when using log-barriers, as raising the barrier raises the value of the loss over time.) 

The major difference (and this is really a hack), is that I regroup all 2D slices from a single 3D volume into the same batch. While this does not affect the loss computations, this allows me to directly compute the 3D dice (what we are ultimately interested in). Ideally, I would like to do that as well for the training loop, but its usually not possible due to memory limitations. 

Let me know, happy to help,
Hoel

**Comment by paumarquez** [comment]
Thanks for answering so fast!

Thanks for the validation/training hack too, I didn't notice that and it's very useful.

However, I did not mean that. My question is that when computing the bounds, they are calculated given the whole image (as you said), so the total positive amount of area is the sum of the area for all the bounding boxes in the image.

Instead, when doing the forward through the network, only a part of the image is given to the model. This means that a mask of the image crop is being predicted while you are comparing the total positive area of that slice prediction with the bounds of the full image, instead of the bounds for that slice.

Am I understanding it wrong or is it suppose to be this way?

**Comment by HKervadec** [comment]
Ok. Sorry for the delayed reply.

> However, I did not mean that. My question is that when computing the bounds, they are calculated given the whole image (as you said), so the total positive amount of area is the sum of the area for all the bounding boxes in the image.

Yes this is correct

> Instead, when doing the forward through the network, only a part of the image is given to the model.

No, the whole image is given. In the end, I do not use sub-patching. 

I think the confusion (my fault really), comes from the left-over code, supposed to handle sub-patching (that would be useful in 3D). I.e., https://github.com/LIVIAETS/boxes_tightness_prior/blob/master/main.py#L147 
I have removed all that in another repository (of the boundary loss), but I haven't done that in this repository. 

I have some plans in the future, to merge the multiple repository that I have into a single one (too keep up with new developments), and I think I'll completely remove that part – it is only adding confusion and reducing readability.

Let me know,

Hoel

**Comment by paumarquez** [comment]
I had do use sub-patching for my 2D problem anyway, since I was having memory issues.

Thanks a lot for your implementation, I have extracted the losses from your code and implemented in mine and it turned out to work pretty good, even though I did not have too much time to experiment with it.

I tried to isolate the losses with their constraints so that it was easier to extract and move the modules to another project for the next one who wants to implement these weakly supervised techniques.

You can check it out [here](https://github.com/paumarquez/AI4Agriculture-grape-detection) if you are curious. In the report you can see your model in action for grape detection!

Thanks!

**Comment by HKervadec** [comment]
> In the report you can see your model **in action for grape detection**!

Hahaha, I gotta say that I did not see that one coming. That is a really cool application. 

I have skimmed through your notebooks, good job! Feel free to share your thesis/report once it is done, I will happily check it. 

For now I will close this issue (as it seems you successfully managed to integrate what you needed). But feel free to come back at it, or open another one should you have any question.

Best,

Hoel
